PyTorch GPU CI - `keras/layers/reshaping/flatten_test.py`

FAILED keras/layers/reshaping/flatten_test.py::FlattenTest::test_flatten_dense

```
________________________ FlattenTest.test_flatten_dense ________________________

element = (array([[[[0.        , 0.        , 0.        , 0.        , 0.        ],
         [0.91335386, 0.9621746 , 0.        , ...0, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.8055, 0.9901, 0.0000,
         0.9847, 0.0000, 0.0000]], device='cuda:0'))
element_signature = None

    def normalize_element(element, element_signature=None):
      """Normalizes a nested structure of element components.

      * Components matching `SparseTensorSpec` are converted to `SparseTensor`.
      * Components matching `RaggedTensorSpec` are converted to `RaggedTensor`.
      * Components matching `VariableSpec` are converted to `Tensor`.
      * Components matching `DatasetSpec` or `TensorArraySpec` are passed through.
      * `CompositeTensor` components are passed through.
      * All other components are converted to `Tensor`.

      Args:
        element: A nested structure of individual components.
        element_signature: (Optional.) A nested structure of `tf.DType` objects
          corresponding to each component of `element`. If specified, it will be
          used to set the exact type of output tensor when converting input
          components which are not tensors themselves (e.g. numpy arrays, native
          python types, etc.)

      Returns:
        A nested structure of `Tensor`, `Variable`, `Dataset`, `SparseTensor`,
        `RaggedTensor`, or `TensorArray` objects.
      """
      normalized_components = []
      if element_signature is None:
        components = nest.flatten(element)
        flattened_signature = [None] * len(components)
        pack_as = element
      else:
        flattened_signature = nest.flatten(element_signature)
        components = nest.flatten_up_to(element_signature, element)
        pack_as = element_signature
      with ops.name_scope("normalize_element"):
        for i, (t, spec) in enumerate(zip(components, flattened_signature)):
          try:
            if spec is None:
>             spec = type_spec_from_value(t, use_fallback=False)

venv2/lib/python3.9/site-packages/tensorflow/python/data/util/structure.py:105:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

element = tensor([[0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.9134, 0.9622, 0.0000, 0.8611,
         0.0000, 0.0000, 0.0000, 0.00...00, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.8055, 0.9901, 0.0000,
         0.9847, 0.0000, 0.0000]], device='cuda:0')
use_fallback = False

    def type_spec_from_value(element, use_fallback=True):
      """Creates a type specification for the given value.

      Args:
        element: The element to create the type specification for.
        use_fallback: Whether to fall back to converting the element to a tensor
          in order to compute its `TypeSpec`.

      Returns:
        A nested structure of `TypeSpec`s that represents the type specification
        of `element`.

      Raises:
        TypeError: If a `TypeSpec` cannot be built for `element`, because its type
          is not supported.
      """
      spec = type_spec._type_spec_from_value(element)  # pylint: disable=protected-access
      if spec is not None:
        return spec

      if isinstance(element, collections_abc.Mapping):
        # We create a shallow copy in an attempt to preserve the key order.
        #
        # Note that we do not guarantee that the key order is preserved, which is
        # a limitation inherited from `copy()`. As a consequence, callers of
        # `type_spec_from_value` should not assume that the key order of a `dict`
        # in the returned nested structure matches the key order of the
        # corresponding `dict` in the input value.
        if isinstance(element, collections.defaultdict):
          ctor = lambda items: type(element)(element.default_factory, items)
        else:
          ctor = type(element)
        return ctor([(k, type_spec_from_value(v)) for k, v in element.items()])

      if isinstance(element, tuple):
        if hasattr(element, "_fields") and isinstance(
            element._fields, collections_abc.Sequence) and all(
                isinstance(f, str) for f in element._fields):
          if isinstance(element, wrapt.ObjectProxy):
            element_type = type(element.__wrapped__)
          else:
            element_type = type(element)
          # `element` is a namedtuple
          return element_type(*[type_spec_from_value(v) for v in element])
        # `element` is not a namedtuple
        return tuple([type_spec_from_value(v) for v in element])

      if hasattr(element.__class__, "__attrs_attrs__"):
        # `element` is an `attr.s` decorated class
        attrs = getattr(element.__class__, "__attrs_attrs__")
        return type(element)(*[
            type_spec_from_value(getattr(element, a.name)) for a in attrs
        ])

      if isinstance(element, CustomNestProtocol):
        # pylint: disable=protected-access
        metadata, children = element.__tf_flatten__()
        return element.__tf_unflatten__(metadata, type_spec_from_value(children))
        # pylint: enable=protected-access

      if use_fallback:
        # As a fallback try converting the element to a tensor.
        try:
          tensor = ops.convert_to_tensor(element)
          spec = type_spec_from_value(tensor)
          if spec is not None:
            return spec
        except (ValueError, TypeError) as e:
          logging.vlog(
              3, "Failed to convert %r to tensor: %s" % (type(element).__name__, e))

>     raise TypeError("Could not build a `TypeSpec` for {} with type {}".format(
          element,
          type(element).__name__))
E     TypeError: Could not build a `TypeSpec` for tensor([[0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.9134, 0.9622, 0.0000, 0.8611,
E              0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.8533, 0.0000, 0.8031,
E              0.0000, 0.0000, 0.8321, 0.9154, 0.9725, 0.0000, 0.0000, 0.0000, 0.0000,
E              0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
E              0.0000, 0.0000, 0.9188, 0.0000, 0.0000, 0.8277, 0.0000, 0.0000, 0.8619,
E              0.0000, 0.8398, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
E              0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.8648, 0.0000, 0.0000, 0.0000,
E              0.0000, 0.0000, 0.8619, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
E              0.0000, 0.9976, 0.0000],
E             [0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.9847, 0.0000, 0.0000,
E              0.0000, 0.0000, 0.9269, 0.0000, 0.0000, 0.9879, 0.0000, 0.0000, 0.0000,
E              0.8812, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.9133, 0.0000,
E              0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.8203, 0.0000,
E              0.0000, 0.8743, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
E              0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.9185, 0.0000,
E              0.0000, 0.0000, 0.0000, 0.9712, 0.0000, 0.9633, 0.0000, 0.9411, 0.0000,
E              0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.8071, 0.0000, 0.0000, 0.8566,
E              0.0000, 0.0000, 0.8047],
E             [0.9898, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
E              0.0000, 0.0000, 0.8284, 0.0000, 0.9270, 0.0000, 0.0000, 0.0000, 0.0000,
E              0.9252, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.8389, 0.0000,
E              0.0000, 0.8965, 0.0000, 0.0000, 0.9717, 0.8453, 0.0000, 0.8955, 0.0000,
E              0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
E              0.0000, 0.0000, 0.9891, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
E              0.9250, 0.0000, 0.0000, 0.8943, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
E              0.8193, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
E              0.0000, 0.0000, 0.8368],
E             [0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.9275, 0.0000, 0.0000, 0.0000,
E              0.0000, 0.0000, 0.0000, 0.9983, 0.9145, 0.9813, 0.9295, 0.0000, 0.0000,
E              0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.8429, 0.0000, 0.0000, 0.0000,
E              0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
E              0.0000, 0.0000, 0.0000, 0.8853, 0.0000, 0.0000, 0.0000, 0.0000, 0.9291,
E              0.0000, 0.0000, 0.8491, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
E              0.0000, 0.0000, 0.0000, 0.0000, 0.9940, 0.9922, 0.0000, 0.0000, 0.0000,
E              0.8829, 0.8066, 0.0000, 0.8760, 0.0000, 0.0000, 0.0000, 0.0000, 0.8761,
E              0.0000, 0.0000, 0.8248],
E             [0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.8337,
E              0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
E              0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.9782, 0.8596, 0.0000,
E              0.0000, 0.9672, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.9124,
E              0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.8983, 0.0000, 0.0000, 0.0000,
E              0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.8635, 0.9678, 0.0000,
E              0.9455, 0.0000, 0.0000, 0.9968, 0.0000, 0.0000, 0.0000, 0.0000, 0.9962,
E              0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.8896, 0.0000,
E              0.9485, 0.0000, 0.0000],
E             [0.8484, 0.8188, 0.9489, 0.0000, 0.0000, 0.0000, 0.0000, 0.8250, 0.0000,
E              0.0000, 0.0000, 0.0000, 0.8984, 0.9229, 0.0000, 0.0000, 0.8317, 0.9740,
E              0.0000, 0.0000, 0.8739, 0.0000, 0.0000, 0.0000, 0.9351, 0.9442, 0.0000,
E              0.0000, 0.8209, 0.0000, 0.0000, 0.8631, 0.0000, 0.9192, 0.0000, 0.0000,
E              0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
E              0.0000, 0.0000, 0.8126, 0.0000, 0.8815, 0.0000, 0.0000, 0.9195, 0.0000,
E              0.0000, 0.0000, 0.0000, 0.0000, 0.8043, 0.0000, 0.9461, 0.9800, 0.0000,
E              0.9124, 0.0000, 0.0000, 0.0000, 0.8089, 0.0000, 0.0000, 0.8280, 0.0000,
E              0.0000, 0.0000, 0.0000],
E             [0.0000, 0.0000, 0.0000, 0.0000, 0.9957, 0.0000, 0.9711, 0.0000, 0.0000,
E              0.8173, 0.0000, 0.0000, 0.0000, 0.8747, 0.0000, 0.0000, 0.0000, 0.0000,
E              0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.8371, 0.0000, 0.0000,
E              0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.8039,
E              0.8137, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
E              0.0000, 0.0000, 0.8339, 0.0000, 0.0000, 0.8951, 0.0000, 0.0000, 0.0000,
E              0.8747, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
E              0.0000, 0.0000, 0.0000, 0.0000, 0.9863, 0.0000, 0.0000, 0.9665, 0.0000,
E              0.0000, 0.0000, 0.0000],
E             [0.0000, 0.0000, 0.8758, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
E              0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.9356, 0.0000, 0.0000, 0.0000,
E              0.0000, 0.0000, 0.9141, 0.0000, 0.8546, 0.0000, 0.0000, 0.0000, 0.0000,
E              0.0000, 0.9149, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
E              0.0000, 0.9058, 0.0000, 0.8235, 0.0000, 0.0000, 0.0000, 0.0000, 0.8478,
E              0.8969, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
E              0.9875, 0.0000, 0.9137, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
E              0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.9199,
E              0.8422, 0.0000, 0.0000],
E             [0.0000, 0.0000, 0.9096, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
E              0.0000, 0.9050, 0.0000, 0.0000, 0.8087, 0.0000, 0.0000, 0.0000, 0.0000,
E              0.0000, 0.0000, 0.0000, 0.0000, 0.9416, 0.0000, 0.0000, 0.0000, 0.0000,
E              0.0000, 0.0000, 0.9527, 0.8264, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
E              0.0000, 0.0000, 0.0000, 0.8276, 0.0000, 0.0000, 0.8982, 0.0000, 0.9149,
E              0.8103, 0.0000, 0.0000, 0.0000, 0.0000, 0.8465, 0.0000, 0.0000, 0.0000,
E              0.0000, 0.8434, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
E              0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
E              0.0000, 0.0000, 0.0000],
E             [0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.8848, 0.9251,
E              0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.9005,
E              0.9753, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.9752,
E              0.0000, 0.0000, 0.0000, 0.0000, 0.8263, 0.0000, 0.8952, 0.0000, 0.0000,
E              0.0000, 0.8946, 0.8337, 0.0000, 0.9399, 0.0000, 0.0000, 0.0000, 0.0000,
E              0.8611, 0.0000, 0.9642, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
E              0.9326, 0.0000, 0.0000, 0.0000, 0.0000, 0.8563, 0.0000, 0.9573, 0.0000,
E              0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.8055, 0.9901, 0.0000,
E              0.9847, 0.0000, 0.0000]], device='cuda:0') with type Tensor

venv2/lib/python3.9/site-packages/tensorflow/python/data/util/structure.py:514: TypeError

During handling of the above exception, another exception occurred:

self = <keras.layers.reshaping.flatten_test.FlattenTest testMethod=test_flatten_dense>
sparse = False

    @parameterized.named_parameters(
        [
            {"testcase_name": "dense", "sparse": False},
            {"testcase_name": "sparse", "sparse": True},
        ]
    )
    @pytest.mark.requires_trainable_backend
    def test_flatten(self, sparse):
        if sparse and not backend.SUPPORTS_SPARSE_TENSORS:
            pytest.skip("Backend does not support sparse tensors.")

        inputs = np.random.random((10, 3, 5, 5)).astype("float32")
        # Make the ndarray relatively sparse
        inputs = np.multiply(inputs, inputs >= 0.8)
        expected_output_channels_last = ops.convert_to_tensor(
            np.reshape(inputs, (-1, 5 * 5 * 3))
        )
        expected_output_channels_first = ops.convert_to_tensor(
            np.reshape(np.transpose(inputs, (0, 2, 3, 1)), (-1, 5 * 5 * 3))
        )
        if sparse:
            import tensorflow as tf

            inputs = tf.sparse.from_dense(inputs)
            expected_output_channels_last = tf.sparse.from_dense(
                expected_output_channels_last
            )
            expected_output_channels_first = tf.sparse.from_dense(
                expected_output_channels_first
            )

        # Test default data_format and channels_last
>       self.run_layer_test(
            layers.Flatten,
            init_kwargs={},
            input_data=inputs,
            input_sparse=True,
            expected_output=expected_output_channels_last,
            expected_output_sparse=sparse,
            run_training_check=not sparse,
        )

keras/layers/reshaping/flatten_test.py:44:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
keras/testing/test_case.py:384: in run_layer_test
    run_training_step(layer, input_data, output_data)
keras/testing/test_case.py:314: in run_training_step
    dataset = tf.data.Dataset.from_tensors(
venv2/lib/python3.9/site-packages/tensorflow/python/data/ops/dataset_ops.py:740: in from_tensors
    return from_tensors_op._from_tensors(tensors, name)
venv2/lib/python3.9/site-packages/tensorflow/python/data/ops/from_tensors_op.py:23: in _from_tensors
    return _TensorDataset(tensors, name)
venv2/lib/python3.9/site-packages/tensorflow/python/data/ops/from_tensors_op.py:31: in __init__
    element = structure.normalize_element(element)
venv2/lib/python3.9/site-packages/tensorflow/python/data/util/structure.py:110: in normalize_element
    ops.convert_to_tensor(t, name="component_%d" % i))
venv2/lib/python3.9/site-packages/tensorflow/python/profiler/trace.py:183: in wrapped
    return func(*args, **kwargs)
venv2/lib/python3.9/site-packages/tensorflow/python/framework/ops.py:696: in convert_to_tensor
    return tensor_conversion_registry.convert(
venv2/lib/python3.9/site-packages/tensorflow/python/framework/tensor_conversion_registry.py:234: in convert
    ret = conversion_func(value, dtype=dtype, name=name, as_ref=as_ref)
venv2/lib/python3.9/site-packages/tensorflow/python/framework/constant_op.py:335: in _constant_tensor_conversion_function
    return constant(v, dtype=dtype, name=name)
venv2/lib/python3.9/site-packages/tensorflow/python/ops/weak_tensor_ops.py:142: in wrapper
    return op(*args, **kwargs)
venv2/lib/python3.9/site-packages/tensorflow/python/framework/constant_op.py:271: in constant
    return _constant_impl(value, dtype, shape, name, verify_shape=False,
venv2/lib/python3.9/site-packages/tensorflow/python/framework/constant_op.py:284: in _constant_impl
    return _constant_eager_impl(ctx, value, dtype, shape, verify_shape)
venv2/lib/python3.9/site-packages/tensorflow/python/framework/constant_op.py:296: in _constant_eager_impl
    t = convert_to_eager_tensor(value, ctx, dtype)
venv2/lib/python3.9/site-packages/tensorflow/python/framework/constant_op.py:103: in convert_to_eager_tensor
    return ops.EagerTensor(value, ctx.device_name, dtype)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = tensor([[0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.9134, 0.9622, 0.0000, 0.8611,
         0.0000, 0.0000, 0.0000, 0.00...00, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.8055, 0.9901, 0.0000,
         0.9847, 0.0000, 0.0000]], device='cuda:0')
dtype = None

    def __array__(self, dtype=None):
        if has_torch_function_unary(self):
            return handle_torch_function(Tensor.__array__, (self,), self, dtype=dtype)
        if dtype is None:
>           return self.numpy()
E           TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.

venv2/lib/python3.9/site-packages/torch/_tensor.py:1030: TypeError
=========================== short test summary info ============================
FAILED keras/layers/reshaping/flatten_test.py::FlattenTest::test_flatten_dense
```

--- a/keras/kokoro/github/ubuntu/gpu/build.sh
+++ b/keras/kokoro/github/ubuntu/gpu/build.sh
@@ -70,6 +70,5 @@ then
    # TODO: Fix the failing Torch GPU CI tests.
    pytest keras --ignore keras/applications \
                --ignore keras/layers/preprocessing/feature_space_test.py \
-               --ignore keras/layers/reshaping/flatten_test.py \
                --cov=keras
 fi

--- a/keras/testing/test_case.py
+++ b/keras/testing/test_case.py
@@ -309,8 +309,19 @@ class TestCase(unittest.TestCase):
             model = TestModel(layer)
 
             if input_sparse:
+                import tensorflow as tf
+
+                if backend.backend() == "torch":
+                    # Bring output Torch tensors to CPU before using `tf.data`
+                    dataset = tf.data.Dataset.from_tensors(
+                        (input_data, backend.convert_to_numpy(output_data))
+                    )
+                else:
+                    dataset = tf.data.Dataset.from_tensors(
+                        (input_data, output_data)
+                    )
                 model.compile(optimizer="sgd", loss="mse", jit_compile=False)
-                model.fit(input_data, output_data, verbose=0)
+                model.fit(dataset, verbose=0)
             else:
                 input_data = tree.map_structure(
                     lambda x: backend.convert_to_numpy(x), input_data

--- a/keras/testing/test_case.py
+++ b/keras/testing/test_case.py
@@ -309,13 +309,8 @@ class TestCase(unittest.TestCase):
             model = TestModel(layer)
 
             if input_sparse:
-                import tensorflow as tf
-
-                dataset = tf.data.Dataset.from_tensors(
-                    (input_data, output_data)
-                )
                 model.compile(optimizer="sgd", loss="mse", jit_compile=False)
-                model.fit(dataset, verbose=0)
+                model.fit(input_data, output_data, verbose=0)
             else:
                 input_data = tree.map_structure(
                     lambda x: backend.convert_to_numpy(x), input_data

